miniapps/shifted/sbm_solver.cpp

```diff
@@ -0,0 +1,642 @@
+// Copyright (c) 2010-2021, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "marking.hpp"
+#include "sbm_solver.hpp"
+#include "mfem.hpp"
+
+namespace mfem
+{
+
+double ShiftedFunctionCoefficient::Eval(ElementTransformation & T,
+                                        const IntegrationPoint & ip,
+                                        const Vector &D)
+{
+   Vector transip;
+   T.Transform(ip, transip);
+   for (int i = 0; i < D.Size(); i++)
+   {
+      transip(i) += D(i);
+   }
+
+   return Function(transip);
+}
+
+void SBM2DirichletIntegrator::AssembleFaceMatrix(
+   const FiniteElement &el1, const FiniteElement &el2,
+   FaceElementTransformations &Trans, DenseMatrix &elmat)
+{
+   int dim, ndof1, ndof2, ndof, ndoftotal;
+   double w;
+   DenseMatrix temp_elmat;
+
+   dim = el1.GetDim();
+   ndof1 = el1.GetDof();
+   ndof2 = el2.GetDof();
+   ndoftotal = Trans.ElementType == ElementTransformation::BDR_FACE ?
+               ndof1 : ndof1 + ndof2;
+
+   elmat.SetSize(ndoftotal);
+   elmat = 0.0;
+
+   bool elem1f = true;
```

Comment:
This function is hard to follow and could use more documentation. If someone wants to use this miniapp as the basis for doing SBM for a more complicated application, wouldn't they need to understand the details here rather than just using this function? If so, it should be well documented. It's not clear to me whether this is something internal that users won't need to understand.